plugins/image-prioritizer/helper.php

```diff
@@ -77,6 +77,142 @@ function image_prioritizer_register_tag_visitors( OD_Tag_Visitor_Registry $regis
+	$script = file_get_contents( __DIR__ . '/' . $path ); // phpcs:ignore WordPress.WP.AlternativeFunctions.file_get_contents_file_get_contents -- It's a local filesystem path not a remote request.
+
+	if ( false === $script ) {
+		return '';
+	}
+
+	return $script;
+}
+
+/**
+ * Filters the list of Optimization Detective extension module URLs to include the extension for Image Prioritizer.
+ *
+ * @since n.e.x.t
+ *
+ * @param string[]|mixed $extension_module_urls Extension module URLs.
+ * @return string[] Extension module URLs.
+ */
+function image_prioritizer_filter_extension_module_urls( $extension_module_urls ): array {
+	if ( ! is_array( $extension_module_urls ) ) {
+		$extension_module_urls = array();
+	}
+	$extension_module_urls[] = add_query_arg( 'ver', IMAGE_PRIORITIZER_VERSION, plugin_dir_url( __FILE__ ) . image_prioritizer_get_asset_path( 'detect.js' ) );
+	return $extension_module_urls;
+}
+
+/**
+ * Filters additional properties for the element item schema for Optimization Detective.
+ *
+ * @since n.e.x.t
+ *
+ * @param array<string, array{type: string}> $additional_properties Additional properties.
+ * @return array<string, array{type: string}> Additional properties.
+ */
+function image_prioritizer_add_element_item_schema_properties( array $additional_properties ): array {
+	$additional_properties['lcpElementExternalBackgroundImage'] = array(
+		'type'       => 'object',
+		'properties' => array(
+			'url'   => array(
+				'type'      => 'string',
+				'format'    => 'uri',
+				'required'  => true,
+				'maxLength' => 500, // Image URLs can be quite long.
+			),
+			'tag'   => array(
+				'type'      => 'string',
+				'required'  => true,
+				'minLength' => 1,
+				'maxLength' => 25, // The longest HTML tag name is 10 characters (BLOCKQUOTE and FIGCAPTION). This maxLength accounts for possible Custom Elements that are even longer.
+				'pattern'   => '^[a-zA-Z0-9\-]+$', // Technically emoji can be allowed in a custom element's tag name, but this is not supported here.
+			),
+			'id'    => array(
+				'type'      => array( 'string', 'null' ),
+				'maxLength' => 100, // A reasonable upper-bound length for a long ID.
+				'required'  => true,
+			),
+			'class' => array(
+				'type'      => array( 'string', 'null' ),
+				'maxLength' => 500, // There can be a ton of class names on an element.
+				'required'  => true,
+			),
+		),
+	);
+	return $additional_properties;
+}
+
+/**
+ * Validates that the provided background image URL is valid.
+ *
+ * @since n.e.x.t
+ *
+ * @param bool|WP_Error|mixed  $validity   Validity. Valid if true or a WP_Error without any errors, or invalid otherwise.
+ * @param OD_Strict_URL_Metric $url_metric URL Metric, already validated against the JSON Schema.
+ * @return bool|WP_Error Validity. Valid if true or a WP_Error without any errors, or invalid otherwise.
+ */
+function image_prioritizer_filter_store_url_metric_validity( $validity, OD_Strict_URL_Metric $url_metric ) {
```

Comment:
is part of the goal here to validate that the stored path actually returns an image? Is this something that could be checked on the client side?

Reply:
Conversation moved to https://github.com/WordPress/performance/pull/1713/files#r1866642173